qiskit/quantum_info/operators/unitary.py

```diff
@@ -0,0 +1,268 @@
+            else:
+                numpy.matmul(self.matrix, other.matrix, out=self.matrix)
+            return self
+        else:
+            output = copy.deepcopy(self)
+            if front:
+                numpy.matmul(other.matrix, output.matrix, out=output.matrix)
+            else:
+                numpy.matmul(output.matrix, other.matrix, out=output.matrix)
+            return output
+
+    def conjugate(self, inplace=False):
+        """conjugate of unitary
+
+        Args:
+            inplace (bool): whether to do conjugation of this object
+
+        Returns:
+            Unitary: unitary object
+        """
+        if inplace:
+            numpy.conj(self.matrix, out=self.matrix)
+            return self
+        else:
+            output = copy.deepcopy(self)
+            numpy.conj(output.matrix, out=output.matrix)
+            return output
+
+    def adjoint(self, inplace=False):
+        """
+        Adjoint
+
+        Args:
+            inplace (bool): don't create new structure
+
+        Returns:
+            Unitary: transposed unitary
+        """
+        return self.transpose(inplace=inplace).conjugate(inplace=inplace)
+
+    def transpose(self, inplace=False):
+        """
+        tranpose unitary
+
+        Args:
+            inplace (bool): attempt inplace (relies on numpy.transpose)
+
+        Returns:
+            Unitary: transposed unitary
+        """
+        if inplace:
+            self.matrix = self.matrix.transpose()
+            return self
+        else:
+            output = copy.deepcopy(self)
+            output.matrix = self.matrix.transpose()
+            return output
+
+    @property
+    def matrix(self):
+        """
+        Get matrix representation
+
+        Returns:
+            numpy.ndarray: matrix
+
+        Raises:
+            QiskitError: if representation not defined
+        """
+        if self._matrix is not None:
+            return self._matrix
+        else:
+            raise QiskitError("matrix representation not defined")
+
+    def power(self, n, inplace=False):
+        """Return n-th matrix power.
+
+        Args:
+            n (int): integer power
+            inplace (bool): whether to do operation in place.
+
+        Returns:
+            Unitary: Unitary power
+        """
+        if inplace:
+            if n >= 0:
+                self.matrix = numpy.linalg.matrix_power(self.matrix, n)
+            else:
+                self.matrix = numpy.linalg.matrix_power(
+                    self.matrix.T.conj(), n)
+            return self
+        else:
+            dim = self.dimension
+            uni = Unitary(numpy.empty((dim, dim), dtype='complex'), validate=False)
+            if n >= 0:
+                uni.matrix = numpy.linalg.matrix_power(self.matrix, n)
+            else:
+                uni.matrix = numpy.linalg.matrix_power(
+                    self.matrix.T.conj(), n)
+            return uni
+
+    @matrix.setter
+    def matrix(self, mat):
```

Comment:
I would suggest not allow a public setter method here and making it so that changing the matrix directly should always be done by initialising a new object like `Unitary(mat)`

Reply:
That sounds fine. I think there was also some talk of eliminating `inplace` for similar reasons.